prow/cmd/deck/pr_history.go

```diff
@@ -0,0 +1,266 @@
+		link := "https://github.com/kubernetes/test-infra/tree/master/gubernator#gcs-bucket-layout"
+		err = fmt.Errorf("unrecognized directory layout: %s (see %s)", key, link)
+		return
+	}
+	bucketName = match[1]
+	org = match[2]
+	if org == "" {
+		org = defaultOrg
+	} else {
+		org = strings.TrimSuffix(org, "_")
+	}
+	repo = match[3]
+	if repo == "" {
+		repo = defaultRepo
+	} else {
+		repo = strings.TrimSuffix(repo, "/")
+	}
+	prNumber = match[4]
+	return
+}
+
+// listJobBuilds concurrently lists builds for the given job prefixes that have been run on a PR
+func listJobBuilds(bucket *storage.BucketHandle, jobPrefixes []string) []jobBuilds {
+	jobch := make(chan jobBuilds)
+	defer close(jobch)
+	for i, jobPrefix := range jobPrefixes {
+		go func(i int, jobPrefix string) {
+			builds, err := listSubDirs(bucket, jobPrefix)
+			if err != nil {
+				logrus.WithError(err).Warningf("Error getting builds for job %s", jobPrefix)
+			}
+			jobch <- jobBuilds{
+				index:  i,
+				builds: builds,
+			}
+		}(i, jobPrefix)
+	}
+	jobs := make([]jobBuilds, len(jobPrefixes))
+	for range jobPrefixes {
+		job := <-jobch
+		jobs = append(jobs, job)
+	}
+	return jobs
+}
+
+// getPRBuildData concurrently fetches metadata on each build of each job run on a PR
+func getPRBuildData(bucket *storage.BucketHandle, bucketName string, jobs []jobBuilds) []buildData {
+	buildch := make(chan buildData)
+	defer close(buildch)
+	expected := 0
+	for _, job := range jobs {
+		i := job.index
+		for j, buildName := range job.builds {
+			go func(i, j int, buildName string) {
+				build, err := getBuildData(bucket, buildName)
+				if err != nil {
+					logrus.WithError(err).Warningf("build %s information incomplete", buildName)
+				}
+				split := strings.Split(strings.TrimSuffix(buildName, "/"), "/")
+				build.SpyglassLink = path.Join(spyglassPrefix, bucketName, buildName)
+				build.ID = split[len(split)-1]
+				build.jobIndex = i
+				build.index = j
+				buildch <- build
+			}(i, j, buildName)
+			expected++
+		}
+	}
+	builds := []buildData{}
+	for k := 0; k < expected; k++ {
+		build := <-buildch
+		builds = append(builds, build)
+	}
+	return builds
+}
+
+func updateCommitData(commits map[string]*commitData, org, repo, hash string, buildTime time.Time, width int) {
+	const prefixLength = 7
+
+	commit, ok := commits[hash]
+	if !ok {
+		commits[hash] = &commitData{
+			Hash:       hash,
+			HashPrefix: hash,
+		}
+		commit = commits[hash]
+		if len(hash) > prefixLength {
+			commit.HashPrefix = hash[:prefixLength]
+		}
+		if hash != "unknown" {
+			commit.Link = githubCommitLink(org, repo, hash)
+		}
+	}
+	if buildTime.After(commit.latest) {
+		commit.latest = buildTime
+	}
+	if width > commit.MaxWidth {
+		commit.MaxWidth = width
+	}
+}
+
+func getPRHistory(url *url.URL, config *config.Config, gcsClient *storage.Client) (prHistoryTemplate, error) {
+	start := time.Now()
+	template := prHistoryTemplate{}
+
+	key := strings.TrimPrefix(url.Path, "/pr-history/")
```

Comment:
It's kind of arbitrary, do you have something better in mind?

Reply:
Do we know the PJ name or can we use query parameters? I really am not a fan of the (convoluted) GCS layout string and in general not a fan of taking structured data, flattening it to a string, then parsing it back out.

Reply:
In this context there isn't a PJ name to know - we're going to fetch all the jobs that executed for a given PR from GCS.

Given that these URLs are public-facing, I would prefer to avoid a series of query parameters.

That said, I wonder if it's viable to go the other way: given a path like `/pr-history/[org]/[repo]/[pr]` or `/pr-history/[repo]/[pr]`, figure out where the bucket should be. This would probably make it significantly easier to generate a URL to this page, [match gubernator's approach](https://gubernator.k8s.io/pr/test-infra/10169), and be cleaner for users.

The key difficulty with that approach is that it requires you to know the correct bucket for an org/repo, which may not necessarily be the default. This isn't a problem kubernetes has today (since `kubernetes-security` is not publicly visible and AFAIK everything else this prow serves is in the default bucket), but it is a possible configuration and I don't think we have a good way to do a lookup from orgs/repos to buckets.

I personally think we should either punt making the URLs better or punt supporting non-default buckets to some future time, depending on whether we think anyone actually uses non-default buckets in a single prow and expects them to be publicly viewable.

Reply:
> Given that these URLs are public-facing, I would prefer to avoid a series of query parameters.

What? I seriously doubt anyone is going to be entering these by hand. They're clicking links in a GUI. I don't think we should  care about this.

> This would probably make it significantly easier to generate a URL to this page, match gubernator's approach, and be cleaner for users.

The most technically correct approach here is to ingest the org, repo, and PR. Use the PR to determine the target branch for it, then use the loaded job configuration to determine which jobs would have ran. Use the Plank configuration and the job configuration to determine which GCS buckets those jobs would have pushed to.

>  depending on whether we think anyone actually uses non-default buckets in a single prow and expects them to be publicly viewable.

We do, Gubernator supports it and we should expect this to work.

Reply:
> What? I seriously doubt anyone is going to be entering these by hand. They're clicking links in a GUI. I don't think we should care about this.

I think clean URLs have multiple upsides: they're easier to manipulate (users do do this! sometimes it's easier than clicking through the right maze of links), they're easier to generate, and they're harder to screw up.

That said...

> The most technically correct approach here is to ingest the org, repo, and PR. Use the PR to determine the target branch for it, then use the loaded job configuration to determine which jobs would have ran. Use the Plank configuration and the job configuration to determine which GCS buckets those jobs would have pushed to.

I looked into this, drafted an implementation, and it seems reasonable enough to do. I believe @ibzib is working on taking this approach now. If we do this, we'll instead have URLs like `/pr-history/kubernetes/kubernetes/12345`. Is that acceptable?

> We do, Gubernator supports it and we should expect this to work.

 That's entirely fair.

Reply:
+1 Parsing long strings using regular expressions is admittedly gross and probably error-prone. We should probably change the job history page to also use this strategy, and possibly Spyglass GCS keys as well.

Reply:
Breaking out org and repo and explicitly placing both in the URL is totally reasonable. I just don't want us to perpetuate the optional org/optional repo and underscore nonsense unless we _need_ to.